### PR TITLE
feat: forward parameter-to-sink taint for the lean-walk languages (#1169)

### LINKED ISSUE #1169: feat: forward parameter taint for the lean-walk languages (Go/JS/TS/Java/C#)
Follow-up to #1142 (PR #1167), which delivered forward parameter-to-sink taint for the **Python** walk only.

## Gap
The lean (non-Python) flow walk (`_process_lean_flow` / `_js_call`) does not yet seed parameters as pseudo-origins, so a tainted argument passed to a logging wrapper in Go/JS/TS/Java/C# still stops at the `kind=arg` edge and never reaches the sink inside the callee body.

## Approach
The summary/closure/composition machinery from #1167 is language-agnostic (it keys on function qn + parameter name and composes at finalize). What's missing on the lean side:
- Seed each parameter into the `tainted` map in `_process_lean_flow` (mirror the Python seeding), using each language's parameter-name extraction.
- Record parameter-to-sink in the `_js_call` write-sink branch and the callee-arg branch (mirror the Python `_apply_call` additions).
- `_flow_param_names` already covers Go/JS-TS/C++; positional composition works for those. Java/C# lack recorded parameter names, so positional mapping yields nothing there until that table is extended (matches the existing arg-edge limits) — keyword/direct cases still compose.
- Consider the `_emit_taint_to_sink` path (Rust `println!` macro, C++ `cout` stream) — needs `caller_qn` threaded to record parameter-sink there.

## Tests
Mirror the Python cases from `test_flow_edges.py` in at least one lean-walk language (Go or Java): logging-wrapper positional, two-hop, cross-file, and a negative control.

### PR DESCRIPTION
**Pull request by vitali87**
Closes #1169. Follow-up to #1142 (PR #1167) / #1168 (PR #1170), which delivered forward parameter taint for the **Python** walk only.

## Gap
The lean (non-Python) flow walk never seeded parameters as pseudo-origins, so a tainted argument passed to a logging wrapper in Go/JS/TS/C++ stopped at the `kind=arg` edge and never reached the sink inside the callee body.

## What this does
The summary/closure/composition machinery from #1167/#1168 is language-agnostic (it keys on `(func_qn, param_name)` and composes at finalize). This wires the **lean side** into it:
- **Seed parameters** in `_process_lean_flow`: each parameter becomes a `Taint(params={name})` pseudo-origin, and the positional order is recorded in `_positional_params` (mirrors the Python `process_flow_for_caller`). A new `_lean_parameter_names` dispatch uses the existing per-language extractors (`go_parameter_names`, `js_ts_parameter_names`, `cpp_parameter_names`).
- **Record parameter-to-sink** in `_js_call` (both the write-sink and callee-arg branches), mirroring `_apply_call` — including the per-call-site pass-through token from #1168 so sibling calls don't cross-contaminate.
- **Thread `caller_qn`** into `_emit_taint_to_sink` so the Rust-macro / C++-stream sink paths also record parameter-sinks.

## Coverage
Working positional param-to-sink composition: **Go, JS, TS/TSX, C++** (they have a parameter-name extractor). **Java/C#** parse but have no extractor yet, so their positional composition stays inert (matches the existing arg-edge limits) — documented as the boundary. Direct param-to-return (`return p`) works for free via existing summary carrying; transitive param-to-return on the lean side remains a follow-up.

## Tests
Six Go cases in `test_flow_edges.py` mirroring the Python suite: logging-wrapper positional, two-hop, wrapper-defined-after-caller, cross-file (same package), plus two negative controls (param never sinks; untainted literal argument). The four positive cases were verified **RED without the seeding fix**. Full flow/taint regression: 977 passed, 1 skipped. ruff + ty clean.

## Summary by CodeRabbit

* **New Features**
  * Improved data-flow analysis for Go, JavaScript, TypeScript/TSX, and C++.
  * Added support for positional, reordered, unnamed, destructured, and variadic parameters.
  * Tainted values passed through wrappers, macros, stream sinks, and cross-file flows are now tracked more accurately.

* **Bug Fixes**
  * Reduced missed or incorrectly reported flows involving declaration order and clean arguments.

* **Documentation**
  * Clarified language-specific data-flow and argument-to-return behavior.

### CODE REVIEW

File: codebase_rag/parsers/flow_access/processor.py [hunk 1]
```diff
@@ -123,6 +129,23 @@ def _py_positional_param_names(func_node: Node) -> list[str]:
...
+    # languages have no keyword-call syntax, so every parameter is positional
+    # and this list doubles as the positional-mapping table (mirrors the
+    # Python _py_positional_param_names role). Java/C#/Rust have no extractor
+    # yet -- positional composition stays inert for them until one is added.
+    if language == cs.SupportedLanguage.GO:
+        return go_parameter_names(func_node)
+    if language in cs.JS_TS_LANGUAGES:
+        return js_ts_parameter_names(func_node)
+    if language == cs.SupportedLanguage.CPP:
+        return cpp_parameter_names(func_node)
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/flow_access/processor.py]: Fixed in 057c264e. Replaced the compacting `_lean_parameter_names` with position-aligned `_lean_parameter_slots`: `go_positional_parameter_slots` / `js_ts_positional_parameter_slots` / `cpp_positional_parameter_slots` in `utils.py` now emit **one entry per formal slot**, `None` where the slot binds no simple name (unnamed C++/Go parameter, JS/TS destructuring/default patterns handled explicitly), so `_param_name_for_via` never shifts `arg:<index>` onto the wrong parameter. Variadic/rest metadata is recorded too: `_variadic_params[qn]` holds the variadic slot index and every argument at or after it maps to that parameter. Regression coverage added: JS destructuring, Go variadic, C++ unnamed leading/trailing, plus a positional-reordering contract test. The three shift/variadic cases were verified RED without the fix.

File: codebase_rag/parsers/flow_access/processor.py [hunk 2]
```diff
@@ -572,6 +595,17 @@ def _process_lean_flow(
...
+        # reaches becomes a parameter-taint summary composed at finalize, exactly
+        # as the Python walk does (issue #1169 extends #1142/#1168 to the lean
+        # walk). The composition machinery in finalize is language-agnostic; only
+        # languages with a parameter-name extractor (Go/JS/TS/C++) get names, so
+        # the rest are seeded with nothing and are unaffected.
+        lean_params = _lean_parameter_names(caller_node, ctx.language)
+        for pname in lean_params:
+            tainted[pname] = Taint(frozenset(), frozenset(), frozenset({pname}))
+        if lean_params:
+            self._positional_params[ctx.caller_qn] = lean_params
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/flow_access/processor.py]: Fixed in 057c264e. The extractors no longer compact: each formal slot keeps its index (unnamed Go/C++ params and JS/TS destructuring/default patterns become explicit `None` slots), so a tainted argument in an unnamed leading position maps to `None` (no edge) rather than shifting a later sink-bearing parameter into its place — eliminating the false source-to-sink edges — and a real argument past an unnamed slot keeps its true index (no dropped flows). Variadic slots are mapped so trailing args bind to the `...` parameter. Added a C++ unnamed-leading-param regression that is RED on the old compacting code (false ENV->STDOUT edge) and green now.

File: codebase_rag/tests/test_flow_edges.py [hunk 3]
```diff
@@ -792,6 +792,163 @@ def test_param_taint_passthrough_calls_do_not_cross_contaminate(
...
+            "package main\n\n"
+            'import (\n\t"fmt"\n\t"os"\n)\n\n'
+            "func inner(x string) {\n\tfmt.Println(x)\n}\n\n"
+            "func logIt(msg string) {\n\tinner(msg)\n}\n\n"
+            "func caller() {\n"
+            '\tsecret := os.Getenv("SECRET")\n'
+            "\tlogIt(secret)\n}\n"
+        )
+    }
+    assert _has_go_secret_to_stdout_flow(_run_flow(tmp_path, files))
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/tests/test_flow_edges.py:836]: Added in 057c264e. `test_param_taint_go_positional_reordering` has an outer wrapper forward its parameters to an inner callee in **swapped** order (`logIt(x, y)` calls `inner(y, x)`) where only `inner`'s second parameter sinks, so the secret only reaches the sink if `arg:<index>` maps to the correct slot at every hop. Also added Go variadic, C++ unnamed leading/trailing, and JS destructuring cases exercising the multi-parameter mapping contract.

**Comment by vitali87** [comment]
@coderabbitai the positional-alignment findings are addressed in 057c264e (per-slot mapping with `None` for unnamed/destructured slots, variadic index mapping, and reordering/edge-case regression tests). Please re-review.

**Comment by vitali87** [comment]
@greptileai the positional mapping shift (P1) is fixed in 057c264e — slots stay index-aligned with explicit `None` for unnamed/destructured parameters and variadic mapping for trailing args. Please re-review.

File: codebase_rag/parsers/utils.py [hunk 4]
```diff
@@ -660,6 +660,109 @@ def go_parameter_names(func_node: Node) -> list[str]:
...
+) -> tuple[list[str | None], int | None]:
+    names: list[str | None] = []
+    variadic_index: int | None = None
+    params = func_node.child_by_field_name(cs.FIELD_PARAMETERS)
+    if params is not None:
+        for child in params.named_children:
+            name, is_variadic = _js_ts_parameter_slot(child)
+            if is_variadic and variadic_index is None:
+                variadic_index = len(names)
+            names.append(name)
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/utils.py]: Good catch — both fixed in 2d292dd5. `_js_ts_parameter_slot` now unwraps a typed parameter (`required_parameter`/`optional_parameter`) recursively into its `pattern`, so a typed rest `(...args: string[])` reaches the `rest_pattern` branch and is marked variadic, and a typed `this` reaches the `this` branch and returns None (no slot). `js_ts_positional_parameter_slots` skips `None` results, so a `this` parameter takes no positional slot and does not shift later runtime parameters. Added tests: unit `test_ts_slots_typed_rest_parameter_is_variadic` (`(["args"], 0)`) and `test_ts_slots_this_parameter_takes_no_slot` (`(["input"], None)`), plus an integration test `test_param_taint_ts_this_parameter_does_not_shift` that is RED on the pre-fix code (the real FILE->STDOUT flow was dropped) and green now.

**Comment by vitali87** [comment]
@coderabbitai the typed-rest and TypeScript `this` parameter handling is fixed in 2d292dd5 (recursive typed-pattern unwrap; `this` occupies no positional slot), with unit and integration tests. Please re-review.

File: codebase_rag/parsers/utils.py [hunk 5]
```diff
@@ -660,6 +660,109 @@ def go_parameter_names(func_node: Node) -> list[str]:
...
+
+def _js_ts_parameter_slot(child: Node) -> tuple[str | None, bool]:
+    node_type = child.type
+    if node_type == cs.TS_IDENTIFIER:
+        return safe_decode_text(child), False
+    if node_type in _JS_TS_TYPED_PARAMETERS:
+        pattern = child.child_by_field_name(cs.TS_FIELD_PATTERN)
+        if pattern is not None and pattern.type == cs.TS_IDENTIFIER:
+            return safe_decode_text(pattern), False
+        return None, False
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/utils.py]: Already fixed in commit 2d292dd5 (this review ran on the earlier `b2df5ec7`). `_js_ts_parameter_slot` now unwraps a typed parameter recursively into its `pattern`, so `required_parameter` > `rest_pattern` reaches the rest-pattern branch and returns `(name, True)`. On current HEAD, `function f(prefix: string, ...values: string[])` yields slots `(["prefix", "values"], 1)` — exactly the fixed behaviour your T-Rex harness predicted — so trailing arguments bind to `values` and the source-to-sink flow is emitted. Covered by `test_ts_slots_typed_rest_parameter_is_variadic` (asserts `(["args"], 0)`).

**Comment by vitali87** [comment]
@greptileai the typed-rest variadic finding was already resolved in 2d292dd5 (recursive typed-pattern unwrap; current HEAD yields `(["prefix", "values"], 1)`). Please re-review the latest commit.

File: codebase_rag/parsers/utils.py [hunk 6]
```diff
@@ -660,6 +660,122 @@ def go_parameter_names(func_node: Node) -> list[str]:
...
+            slot = _js_ts_parameter_slot(child)
+            if slot is None:
+                # A TypeScript `this` parameter is not a runtime argument, so it
+                # takes no slot and must not shift the parameters after it.
+                continue
+            name, is_variadic = slot
+            if is_variadic and variadic_index is None:
+                variadic_index = len(names)
+            names.append(name)
+        return names, variadic_index
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/utils.py:753]: This is correct on the current HEAD. `_js_ts_parameter_slot` unwraps the typed pattern recursively, so `...values: string[]` reaches the `rest_pattern` branch and returns `(name, True)` — `function f(prefix: string, ...values: string[])` yields slots `(["prefix", "values"], 1)`. Added an end-to-end regression in aed5c3e1 (`test_param_taint_ts_typed_rest_binds_trailing_args`): a wrapper `logIt(prefix, ...vals)` that sinks `vals`, called with the file read as the trailing argument, emits `resource::FILE::cfg.txt -> resource::STDOUT::<dynamic>`. It passes on HEAD. If T-Rex still reports `env_to_stdout=False`, it is running an earlier commit than aed5c3e1.

File: codebase_rag/parsers/utils.py [hunk 7]
```diff
@@ -660,6 +660,122 @@ def go_parameter_names(func_node: Node) -> list[str]:
...
+    params = func_declarator.child_by_field_name(cs.KEY_PARAMETERS)
+    if params is None:
+        return [], None
+    names: list[str | None] = []
+    for declaration in params.named_children:
+        if declaration.type not in _CPP_PARAMETER_DECLARATIONS:
+            continue
+        param_declarator = declaration.child_by_field_name(cs.FIELD_DECLARATOR)
+        names.append(cpp_declarator_name(param_declarator))
+    return names, None
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/utils.py:776]: This is already correct on the current HEAD. `cpp_positional_parameter_slots` keeps one entry per formal slot with `None` for an unnamed parameter, so `void f(const char*, const char* msg)` yields `[None, "msg"]` and the index never shifts. Two integration tests already lock this in: `test_param_taint_cpp_unnamed_slot_does_not_shift` (secret into the unnamed arg:0 -> no false ENV->STDOUT edge) and `test_param_taint_cpp_unnamed_leading_param_recovers_flow` (secret into the named arg:1 -> the real flow is emitted). Both were verified RED on the pre-fix compacting code and pass now.

**Comment by vitali87** [comment]
@greptileai both P1s are resolved on the latest commit aed5c3e1: C++ unnamed slots keep their position (two RED-verified integration tests), and typed rest binds trailing args (new end-to-end test emits `FILE::cfg.txt -> STDOUT`). Please re-review the latest commit.

File: codebase_rag/parsers/utils.py [hunk 8]
```diff
@@ -660,6 +660,122 @@ def go_parameter_names(func_node: Node) -> list[str]:
...
+                variadic_index = len(names)
+            names.append(name)
+        return names, variadic_index
+    single = func_node.child_by_field_name(cs.TS_FIELD_PARAMETER)
+    if single is not None and single.type == cs.TS_IDENTIFIER:
+        names.append(safe_decode_text(single))
+    return names, variadic_index
+
+
+def cpp_positional_parameter_slots(
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/parsers/utils.py:760]: Verified correct end-to-end on the current HEAD (aed5c3e1). `cpp_positional_parameter_slots` keeps one entry per formal slot: `void logIt(const char*, const char* msg)` yields `([None, "msg"], None)`. Running the full GraphUpdater on the exact scenario:

- secret into the UNNAMED arg:0 (`logIt(s, "safe")`) → **no** `ENV::SECRET -> STDOUT` edge (no false positive)
- secret into the NAMED arg:1 (`logIt("safe", s)`) → the real `ENV::SECRET -> STDOUT` edge IS emitted (no dropped flow)

Both are locked in by `test_param_taint_cpp_unnamed_slot_does_not_shift` and `test_param_taint_cpp_unnamed_leading_param_recovers_flow`, each verified RED on the pre-fix compacting code. The Greptile Review check is green; resolving this thread as addressed.
